yt_dlp/utils.py

```diff
@@ -5576,6 +4970,231 @@ def items_(self):
         return self.__dict__.items()
 
 
+class CaseInsensitiveDict(collections.UserDict):
+    """
+    Store and get items case-insensitively.
+    All keys are assumed to be a string, and are converted to title case.
+    Latter headers are prioritized in constructor
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        for dct in args:
+            if dct is None:
+                continue
+            self.update(dct)
+        self.update(kwargs)
+
+    def __setitem__(self, key, value):
+        super().__setitem__(key.title(), str(value))
+
+    def __getitem__(self, key):
+        return super().__getitem__(key.title())
+
+    def __delitem__(self, key):
+        super().__delitem__(key.title())
```

Comment:
UserDict does operations on the internal dictionary, not itself

Reply:
can u explain?
```py
>>> import collections
>>> class CaseInsensitiveDict(collections.UserDict):
...     def __setitem__(self, key, value):
...         super().__setitem__(key.title(), str(value))
...
...     def __getitem__(self, key):
...         return super().__getitem__(key.title())
...
...     def __delitem__(self, key):
...         super().__delitem__(key.title())
...
...     def __contains__(self, key):
...         if not isinstance(key, str):
...             return False
...         return super().__contains__(key.title())
...
>>>
>>> h = CaseInsensitiveDict(Abcd=1, efGh=2)
>>> print(h)
{'Abcd': '1', 'Efgh': '2'}
>>>
>>> assert h == CaseInsensitiveDict({'ABCD': 1, 'EFGH': 2})  # eq
>>> assert 'ABCD' in h  # contains
>>>
>>> h2 = h.copy()  # copy
>>> h2['AbCd'] = 3
>>> assert h['abcd'] == '1'
>>>
>>> assert h | h2 == h2  # or
>>>
>>> h.update({'aBCd': 3})  # update
>>> assert h == h2
```

Reply:
?

Reply:
everything seems to work with just these 4 functions

Reply:
Actually before we delve into this, does it make sense for this to be derived from ChainMap? See https://github.com/yt-dlp/yt-dlp/pull/2861#discussion_r904691218

Reply:
- copy() doesn't seem to be needed, looks like there is handling for keeping it the same type
- I override __or__ etc. so it can do an OR between a CaseInsensitiveDict and some other dict case-insensitively. Not sure if we want that behavior
- `add` is just a useful method for joining multiple headers of the same key

Reply:
> I override `__or__` etc. so it can do an OR between a CaseInsensitiveDict and some other dict case-insensitively. Not sure if we want that behavior

I sill don't get it. Using my definition:
```py
>>> CaseInsensitiveDict(Abcd=1, efGh=2) == {'Abcd': '1', 'Efgh': '2'}
True
>>> {'Abcd': '1', 'Efgh': '2'} == CaseInsensitiveDict({'ABCD': 1, 'EFGH': 2})
True
>>> CaseInsensitiveDict(Abcd=1, efGh=2) | {'ABCD': 3}
{'Abcd': '3', 'Efgh': '2'}
```

> I override `__init__` so we can pass multiple dicts (similar to merge_headers)

which is why I think it should be a chainmap

> add is just a useful method for joining multiple headers of the same key

This may be problematic if we want to use the class for other things. How useful is this feature?

Reply:
> > I override `__or__` etc. so it can do an OR between a CaseInsensitiveDict and some other dict case-insensitively. Not sure if we want that behavior
> 
> I sill don't get it. Using my definition:
> 
> ```python
> >>> CaseInsensitiveDict(Abcd=1, efGh=2) == {'Abcd': '1', 'Efgh': '2'}
> True
> >>> {'Abcd': '1', 'Efgh': '2'} == CaseInsensitiveDict({'ABCD': 1, 'EFGH': 2})
> True
> >>> CaseInsensitiveDict(Abcd=1, efGh=2) | {'ABCD': 3}
> {'Abcd': '3', 'Efgh': '2'}
> ```

I was just reading through here: https://github.com/python/cpython/blob/3.10/Lib/collections/__init__.py#L1125-L1144

Wouldn't it do OR with our titled() dictionary and whatever the other one is? unless I am misunderstanding something


> > add is just a useful method for joining multiple headers of the same key
> 
> This may be problematic if we want to use the class for other things. How useful is this feature?

true that. It probably isn't really needed now that I think about it.

Reply:
It first `or`'s them as dicts. But when it is cast to `CaseInsensitiveDict`, the duplicate entries are resolved

The whole point of `UserDict` is that it makes inheriting from dict easy. If we had to redefine everything, there is no point :D